File: src/simulation/field/grassField/AgentBrain.java
package simulation.field.grassField;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import simulation.agent.brainPackage.BehaviorManager;
import simulation.common.Node;
import simulation.common.PolarCoordinate;
import simulation.common.State;

//TODO: this class needs to take all the logic of making local decisions away from AcutalRobot
//right now it is not doing anything
public class AgentBrain {

private BehaviorManager behaviorManager;
private boolean isCrashed = false;

private List<Node> newNodes = new ArrayList<Node>();
private List<PolarCoordinate> newPostions = new ArrayList<PolarCoordinate>();

public AgentBrain(BehaviorManager behaviorManager){
	this.behaviorManager = behaviorManager;
}

public void writeDataToRobot(PolarCoordinate[] p, Node[] n) throws IOException, InterruptedException{
	newPostions = Arrays.asList(p);
	newNodes = Arrays.asList(n);
}

public State calcauteNextRobotState(){
	try {
		behaviorManager.updateDeltaPostions(newPostions);
	} catch (IOException e) {
		// TODO Auto-generated catch block
		e.printStackTrace();
	}
	behaviorManager.updateNodeList(newNodes);
	
	isCrashed = checkIfCrashed();
	
	if(isCrashed){
		behaviorManager.setCrashedState();
	}
	
	else {
		behaviorManager.setSwarmState();
	}
	
	return behaviorManager.getCurrentState();
}

private boolean checkIfCrashed() {
	if(!newPostions.isEmpty()){
		if(newPostions.get(0).getName().equals("CRASH")){
		return true;
		}
	}
	return false;
}

public void hasCrashed(){
	isCrashed = true;
}

public void notCrashed(){
	isCrashed = false;
}

public boolean getCrashed(){
	return isCrashed;
}
}

File: src/simulation/agent/brainPackage/AgentBrain.ts
import { BehaviorManager } from './BehaviorManager';
import { Node } from '../../common/Node';
import { PolarCoordinate } from '../../common/PolarCoordinate';
import { State } from '../../common/State';

export class AgentBrain {
  private behaviorManager: BehaviorManager;
  private isCrashed: boolean = false;
  private newNodes: Node[] = [];
  private newPositions: PolarCoordinate[] = [];

  constructor(behaviorManager: BehaviorManager) {
    this.behaviorManager = behaviorManager;
  }

  writeDataToRobot(p: PolarCoordinate[], n: Node[]): void {
    this.newPositions = p;
    this.newNodes = n;
  }

  calculateNextRobotState(): State {
    this.behaviorManager.updateDeltaPositions(this.newPositions);
    this.behaviorManager.updateNodeList(this.newNodes);

    this.isCrashed = this.checkIfCrashed();

    if (this.isCrashed) {
      this.behaviorManager.setCrashedState();
    } else {
      this.behaviorManager.setSwarmState();
    }

    return this.behaviorManager.getCurrentState();
  }

  private checkIfCrashed(): boolean {
    if (this.newPositions.length > 0) {
      if (this.newPositions[0].getName() === 'CRASH') {
        return true;
      }
    }
    return false;
  }

  hasCrashed(): void {
    this.isCrashed = true;
  }

  notCrashed(): void {
    this.isCrashed = false;
  }

  getCrashed(): boolean {
    return this.isCrashed;
  }
}
